feat(test): expect.extend and asymmetric matcher improvements

otgerrogla opened the pull request:
### What does this PR do?

This PR implements the following to `bun test`:
 - `expect.extend({ ... })` support to register custom matchers (#3621)
    - Note: support is still not 100%, it is missing some helpers and variables in the matcher context (i.e. the `this` and `this.utils` inside custom matcher functions)
    - Support for declaring typescript types for custom matchers through declaration merging (same way as jest)
    - Added related tests from Jest
 - `expect.not` support, to negate of asymmetric matchers (e.g. `expect.not.stringContaining('bun')`)
 - Add new asymmetric matchers from Jest (including related tests):
   - `expect.closeTo`
   - `expect.objectContaining`
 - ~Fixes for existing asymmetric matchers:~
   - ~`expect.arrayContaining([])` (with empty array) should match anything, even numbers (that's what jest does)~
 - Adds `expect.resolvedTo` / `expect.rejectedTo`, to support for asynchronous asymmetric matchers (e.g. `expect.resolvedTo.stringContaining('bun')`). This is not in jest, although it existed as a feature proposal: https://github.com/jestjs/jest/issues/10556
 - Some refactoring of the `bun:test` typescript types to accomodate custom matchers, the expect.not, etc.
 - Tests:
   - Tests for all new features and fixes above at `expect.test.js` and `expect-extend.test.js`
   - Improved interop compatibility to run these test files as-is on bun, jest, and vitest
   - Added type-checking to these test files, so they can be used as type-tests for types of `bun:test` (although not automated as of now, just for visual testing)

Implementation notes:
 - ~It adds some capabilities to `generate-classes.ts` (in order to support custom matchers):~
   - ~Option to definine a `getOwnPropertySlot` callback, to dynamically resolve properties at native level.~
   - ~Option to define an `apply` callback, executed when the class instance is called as a function (the class itself, not the constructor).~
 - ~Uses `getOwnPropertySlot` to dynamically resolve custom matchers to improve performance (jest actually populates a new object with all matchers each time)~
 - The custom matchers are currently stored in a singleton object in the ZigGlobalObject

<!-- **Please explain what your changes do**, example: -->

<!--

This adds a new flag --bail to bun test. When set, it will stop running tests after the first failure. This is useful for CI environments where you want to fail fast.

-->

- [X] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [X] Code changes

### How did you verify your code works?

<!-- **For code changes, please include automated tests**. Feel free to uncomment the line below -->

I wrote automated tests (some created + some ported from jest)

<!-- If JavaScript/TypeScript modules or builtins changed:

-->

- [x] I included a test for the new code, or existing tests cover it
- [x] I ran my tests locally and they pass (`bun-debug test test-file-name.test`)

<!-- If Zig files changed:
-->

- [x] I checked the lifetime of memory allocated to verify it's (1) freed and (2) only freed when it should be
- [x] I included a test for the new code, or an existing test covers it
- [x] JSValue used outside outside of the stack is either wrapped in a JSC.Strong or is JSValueProtect'ed
- [x] I wrote TypeScript/JavaScript tests and they pass locally (`bun-debug test test-file-name.test`)


<!-- If new methods, getters, or setters were added to a publicly exposed class:
-->

- [x] I added TypeScript types for the new methods, getters, or setters


<!-- If dependencies in tests changed:

- [ ] I made sure that specific versions of dependencies are used instead of ranged or tagged versions
-->

<!-- If a new builtin ESM/CJS module was added:

- [ ] I updated Aliases in `module_loader.zig` to include the new module
- [ ] I added a test that imports the module
- [ ] I added a test that require() the module
-->

Jarred-Sumner commented:
Thank you very much for this. 

`getOwnPropertySlot` is a neat approach but the downside is it prevents inline-caching, which means that all property accesses will be slow on the `expect` instances (something like 10x - 100x slower). This would have a meaningfully negative performance impact, even when no custom matchers are in use.

There's another way that won't have a performance impact when not in use and will probably be less code. We can directly mutate the prototype used by the `Expect` object. We can define the custom matchers as functions on the prototype. For asymmetric matchers, we can `.get` on that prototype.

otgerrogla commented:
> There's another way that won't have a performance impact when not in use and will probably be less code. We can directly mutate the prototype used by the `Expect` object. We can define the custom matchers as functions on the prototype. For asymmetric matchers, we can `.get` on that prototype.

It's not so direct, because when calling e.g. `expect.myCustomMatcher()` it doesn't directly call the user-provided js matcher function, but calls a native function that wraps the call and applies some logic (parsing and validating the user-provided function return value, etc.)

I guess this can be achieved by using bound native functions. Any other/better idea?

otgerrogla commented:
Actually bound functions won't work, because you lose the reference to the "this" and can't access the matcher flags

otgerrogla commented:
Doesn't work with bound functions either, because you lose the reference to "this" and can't retrieve the flags

Electroid commented on src/bun.js/bindings/bindings.cpp:
```diff
@@ -232,20 +269,24 @@ AsymmetricMatcherResult matchAsymmetricMatcher(JSGlobalObject* globalObject, JSC
-                unsigned expectedLength = expectedArray->length();
-                unsigned otherLength = otherArray->length();
+            if (expectedLength == 0) {
+                // jest makes it so an empty array matches anything,
+                // even a primitive like a string or a number
```
What's an example of this? We have deviated a handfull of times from Jest where something was _obviously_ wrong. (aka. potential to break is outweighed by catching likely errored code)

otgerrogla replied:
I added some tests to `expect.test.js`, and ran them on Jest where they pass
```js
    test("ArrayContaining matches", () => {
      expect(expect.arrayContaining([])).toEqual(42);
      expect(expect.arrayContaining([])).toEqual("jest");
      expect(expect.arrayContaining(["foo"])).toEqual(["foo"]);
      expect(expect.arrayContaining(["foo"])).toEqual(["foo", "bar"]);
      expect(expect.arrayContaining([])).toEqual({});
    });
```
Although I agree this is a very weird behavior

Electroid replied:
Let's change those to fail, so `expect.arrayContaining` fails when not an array. Same with the other asymmetric matchers like `expect.objectContaining`. This just defies common sense, if code breaks from these paths, it was already broken.

Electroid commented on src/bun.js/bindings/bindings.zig:
```diff
@@ -5872,3 +5927,30 @@ pub const ScriptExecutionStatus = enum(i32) {
     stopped = 2,
 };
+
+pub const PropertyAttributes = c_uint;
+
+// ECMA 262-3 8.6.1
+// Property attributes
+pub const PropertyAttribute = struct {
+    pub const ReadOnly: PropertyAttributes = 1 << 1; // property can be only read, not written
+    pub const DontEnum: PropertyAttributes = 1 << 2; // property doesn't appear in (for .. in ..)
+    pub const DontDelete: PropertyAttributes = 1 << 3; // property can't be deleted
+    pub const Accessor: PropertyAttributes = 1 << 4; // property is a getter/setter
+    pub const CustomAccessor: PropertyAttributes = 1 << 5;
+    pub const CustomValue: PropertyAttributes = 1 << 6;
+};
```
This would be a good candidate to use Zig's [packed struct](https://devlog.hexops.com/2022/packed-structs-in-zig/#zig-packed-structs).

otgerrogla replied:
I though about it, but wanted to mimic the C++ definition as close as possible.
Anyways I'm gonna remove this part

Electroid commented on src/bun.js/test/expect.zig:
```diff
@@ -57,8 +69,30 @@ pub const Expect = struct {
         } = .none,
         not: bool = false,
+
+        const BitsetType = i32;
+        const FLAG_NOT: BitsetType = (1 << 0);
+        const FLAG_RESOLVES: BitsetType = (1 << 1);
+        const FLAG_REJECTS: BitsetType = (1 << 2);
+
+        /// Used to pass the flags to C++ asymmetric matchers execution
+        /// Note: keep in sync with the corresponding C++ bitset flags
+        pub fn toBitset(this: Flags) BitsetType {
+            return (if (this.not) FLAG_NOT else 0) | (if (this.promise == .resolves) FLAG_RESOLVES else 0) | (if (this.promise == .rejects) FLAG_REJECTS else 0);
+        }
+
+        pub fn fromBitset(bitset: BitsetType) Flags {
+            return Flags{
+                .not = (bitset & FLAG_NOT) != 0,
+                .promise = if ((bitset & FLAG_RESOLVES) != 0) .resolves else if ((bitset & FLAG_REJECTS) != 0) .rejects else .none,
+            };
+        }
```
Same comment, good candidate for packed structs.

otgerrogla replied:
Indeed, this makes more sense here
Thanks

Electroid commented on src/bun.js/test/expect.zig:
```diff
@@ -3776,6 +4101,98 @@ pub const ExpectStringMatching = struct {
+            return .zero;
+        }
+
+        if (Jest.runner.?.pending_test == null) {
+            const err = globalObject.createErrorInstance("expect.closeTo() must be called in a test", .{});
+            err.put(globalObject, ZigString.static("name"), ZigString.init("TestNotRunningError").toValueGC(globalObject));
+            globalObject.throwValue(err);
+            return .zero;
+        }
```
We should fix this, this technically isn't right. Asymmetric matchers should be able to be defined outside the context of a test. For example:

```js
test.each([
  {
    actual: { a: 1, b: 2 },
    // this is not called in the test scope, but in global scope
    expected: { a: expect.closeTo(1), b: expect.closeTo(2) },
  }
])("%s", ({ actual, expected }) => {
  expect(actual).toEqual(expected);
});
```

I think we can just remove these checks for the matchers.

otgerrogla replied:
I put it in because I saw the existing asymmetric matchers had this check, although I agree this is a good point

otgerrogla replied:
Should I remove it from all asymmetric matchers?

Electroid replied:
Yep

Electroid reviewed (commented):
Very exciting changes. Someone else on the team will do a more through review, but left a few comments

otgerrogla commented:
@Jarred-Sumner I replaced the usage of `getOwnPropertySlot`, so now when `expect.extend` is called, it mutates the prototypes/constructor by adding new `JSValue` instances of `JSFunction` which point to a same native function, but each has a property `fn` with the user-provided matcher function so it can known which custom matcher to use. That's the best way I could came up with to do it while still allowing the matchers to access the flags.

I also moved the flags from JSValues to being fields in the zig structs themselves.

Jarred-Sumner commented on src/bun.js/test/expect.zig:
```diff
@@ -3707,10 +3803,197 @@ pub const Expect = struct {
+        const message_fn = matcher_result.get(globalObject, "message") orelse JSValue.undefined;
+        if (message_fn.isCallable(globalObject.vm())) {
+            var message_result = message_fn.callWithGlobalThis(globalObject, &[_]JSValue{});
+            if (message_result.isAnyError()) {
```
```suggestion
            if (message_result.toError(globalObject)) |err| {
                globalObject.bunVM().onUnhandledError(globalObject, err);
                return false;
```

Jarred-Sumner commented on src/bun.js/test/expect.zig:
```diff
@@ -3707,10 +3803,197 @@ pub const Expect = struct {
+        const args_ptr = callFrame.argumentsPtr();
+        const args_count = callFrame.argumentsCount();
+        var allocator = std.heap.stackFallback(8 * @sizeOf(JSValue), globalObject.allocator());
+        var matcher_args = std.ArrayList(JSValue).initCapacity(allocator.get(), args_count + 1) catch {
```
Can you add a test that calls a custom matcher 10,000 times in a loop?

This current implementation will likely cause GC issues & crashes. JSC's GC stack scanning is limited. I kind of think we might need to implement at least part of this in C++ to avoid that situation. 

In C++, I would use a `ArgsList` here

otgerrogla replied:
Run this test locally several times with 1M calls and didn't crash, I think this is safe as the JSValue arguments are already "pinned" (although I might be wrong):
- For symmetric matchers:
  - First arg (the expected value): stored/owned on the `Expect` instance
  - Rest: pinned on the callFrame
- For asymmetric matchers:
  - First arg (the expected value): comes from a callFrame up above (deepEquals or whatever call)
  - Rest: stored/owned in a JSArray at the `ExpectCustomAsymmetricMatcher` instance
 
Also what difference would using an `ArgList` make? As I got it you still have to allocate an array to put the JSValues, as you need to prepend the expected value for the call

Jarred-Sumner commented on src/bun.js/test/expect.zig:
```diff
@@ -3948,10 +4459,382 @@ pub const ExpectArrayContaining = struct {
+
+        // retrieve the user-provided matcher function
+        var func: JSValue = callFrame.callee();
+        var matcher_fn = func.get(globalObject, Expect.custom_matcher_fn_property) orelse unreachable;
```
We can use a `JSC::JSBoundFunction` here I think. JSBoundFunction can preserve the `this` and any additional arguments. `JSBoundFunction` gets inlined in the JIT, so it has very little overhead.

If we cannot use a JSBoundFunction, then the backup would be to store the custom matcher as a `WriteBarrier` on the instance itself. Then we can avoid the (observable) property access.

otgerrogla replied:
I tried with a bound function but got no way to recover the Expect/ExpectStatic instance to retrieve the flags as you need to specify the `this` on creation (i.e. in the expect.extends execution).

I will do the second way, I thought about subclassing `JSFunction` but didn't know how to go about it, thanks for the tip!

Jarred-Sumner replied:
Another option would be to make custom matchers a `CustomGetterSetter` which returns the JSFunction subclass.

It would work like [EventListenerMap in WebCore](https://github.com/oven-sh/WebKit/blob/0aa1f6dfc9b48be1e89b8f709fd44e3c88feaf33/Source/WebCore/dom/EventListenerMap.h#L81-L89):


```c++
template<typename Visitor>
void EventListenerMap::visitJSEventListeners(Visitor& visitor)
{
    Locker locker { m_lock };
    for (auto& entry : m_entries) {
        for (auto& eventListener : entry.second)
            eventListener->callback().visitJSFunction(visitor);
    }
}
```

The functions would lazily be created on each instance on access. Then, cached and visited within [`visitAdditionalChildren`](https://github.com/oven-sh/WebKit/blob/0aa1f6dfc9b48be1e89b8f709fd44e3c88feaf33/Source/WebCore/bindings/js/JSWorkerGlobalScopeCustom.cpp#L40-L54) on each instance:
```c++
template<typename Visitor>
void JSWorkerGlobalScope::visitAdditionalChildren(Visitor& visitor)
{
    if (auto* location = wrapped().optionalLocation())
        addWebCoreOpaqueRoot(visitor, *location);
    if (auto* navigator = wrapped().optionalNavigator())
        addWebCoreOpaqueRoot(visitor, *navigator);
    ScriptExecutionContext& context = wrapped();
    addWebCoreOpaqueRoot(visitor, context);
    
    // Normally JSEventTargetCustom.cpp's JSEventTarget::visitAdditionalChildren() would call this. But
    // even though WorkerGlobalScope is an EventTarget, JSWorkerGlobalScope does not subclass
    // JSEventTarget, so we need to do this here.
    wrapped().visitJSEventListeners(visitor);
}
```

otgerrogla replied:
Did it with the JSFunction subclassing approach, seems to work fine!
This way you can access both the the expect instance to get the flags (`callframe.this()`) and recover the user-provided matcher function from `calframe.callee()` which is the new subclass instance

dylan-conway commented on test/js/bun/test/expect-extend.test.js:
```diff
@@ -0,0 +1,213 @@
+      default: 42,
+    }),
+  ).toThrow(/*'expect.extend: `default` is not a valid matcher. Must be a function, is "number"'*/);
+  expect(() =>
```
were these meant to be commented? they still pass uncommented

otgerrogla replied:
forgot about them, thanks

dylan-conway commented on src/bun.js/bindings/bindings.cpp:
```diff
@@ -129,23 +129,51 @@ using namespace JSC;
-AsymmetricMatcherResult matchAsymmetricMatcher(JSGlobalObject* globalObject, JSCell* matcherPropCell, JSValue otherProp, ThrowScope* throwScope)
+template<class T>
+bool readFlagsAndProcessPromise(JSValue& instanceValue, T* instance, ExpectFlags& flags, JSGlobalObject* globalObject, JSValue& value)
+{
```
is the template necessary? `instance` doesn't seem to be used

otgerrogla replied:
true! it was a leftover of when the flags where stored as JSValue

dylan-conway commented on src/bun.js/bindings/bindings.cpp:
```diff
@@ -272,11 +309,74 @@ AsymmetricMatcherResult matchAsymmetricMatcher(JSGlobalObject* globalObject, JSC
+{
+    ExpectFlags flags = ExpectFlags();
+    AsymmetricMatcherResult result = matchAsymmetricMatcherAndGetFlags(globalObject, matcherProp, otherProp, throwScope, flags);
+    if (result != AsymmetricMatcherResult::NOT_MATCHER && (flags & FLAG_NOT)) {
```
In jest both `expect.not.any` and `expect.not.anything` are not valid asymmetric matchers. We will need to throw for both of them.
<img width="605" alt="Screenshot 2023-11-28 at 4 14 24 PM" src="https://github.com/oven-sh/bun/assets/35280289/aa0c2f62-bc3d-4d79-8d16-702869a22706">

otgerrogla replied:
I know, but I feel like we can support them "for free" as a feature, as it's not gonna break any existing jest test. They might be useful, e.g.:
- match a null or undefined with `expect.not.anything()`
- match something that's not a Date with `expect.not.any(Date)`

dylan-conway replied:
I agree, let's keep them. `expect.not.any(Class)` seems very useful

dylan-conway commented on src/bun.js/test/pretty_format.zig:
```diff
@@ -1986,4 +1951,141 @@ pub const JestPrettyFormat = struct {
+
+            printAsymmetricMatcherPromisePrefix(matcher.flags, this, writer);
+            if (matcher.flags.not) {
+                this.addForNewLine("Any<".len);
```
`"NotAny<"` and `"Any<"` need to swap places

otgerrogla replied:
good catch!

Jarred-Sumner commented on src/bun.js/bindings/bindings.cpp:
```diff
@@ -1685,6 +1805,19 @@ void JSC__JSFunction__optimizeSoon(JSC__JSValue JSValue0)
+    if (JSC::JSFunction* func = jsDynamicCast<JSC::JSFunction*>(value)) {
+        auto* sourceCode = func->sourceCode();
+        if (sourceCode != nullptr) { // native functions have no source code
+            *outSourceCode = Zig::toZigString(sourceCode->view());
```
This will become invalid memory

It should be a BunString.

Jarred-Sumner commented on src/bun.js/bindings/bindings.cpp:
```diff
@@ -1685,6 +1805,19 @@ void JSC__JSFunction__optimizeSoon(JSC__JSValue JSValue0)
     JSC::optimizeNextInvocation(value);
 }
 
+bool JSC__JSFunction__getSourceCode(JSC__JSValue JSValue0, ZigString* outSourceCode)
```
Do we need to do this? `toString` on a function returns the source code for the function

Jarred-Sumner commented:
Thank you for very much. I'm going to merge this and then make some small tweaks.
